_posts/2019/2019-09-08-use-a-custom-validator.md

```diff
@@ -0,0 +1,81 @@
+---
+title: "Use a custom validator"
+description: "A little structure for repeated checks"
+layout: article
+category: ruby
+image:
+  base: '2019/use-a-custom-validator'
+  alt: 'Dropped Ice Cream'
+  credit: Sarah Kilian
+  source: "https://unsplash.com/photos/52jRtc2S_VE"
+
+---
+
+ActiveModel’s validations, used by calling `validates` inside your model with various options, can be supplemented with your own custom classes. The Rails Guides contains [a section on custom validations](https://guides.rubyonrails.org/active_record_validations.html#custom-validators).
+
+
+## Instead of...
+
+...using complex code in your `validates` calls:
+
+```ruby
+class Invite
+  validates :invitee_email, format: {
+    with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i,
+    message: "is not an email"
+  }
+end
+```
+
+## Use...
+
+...a validator class and extract the logic into it:
+
+```ruby
+class EmailValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    unless value =~ /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i
+      record.errors[attribute] << (options[:message] || "is not an email")
+    end
+  end
+end
+```
+
+```ruby
+class Invite
+  validates :invitee_email, email: true
```

Comment:
I'm guessing there's 'Rails magic' here that works out which validator to use when it sees `email: true` as an option?